### DISCUSSION TITLE
"gramine-sgx helloworld " faild

### DISCUSSION THREAD

**Discussion by jyizheng** [gramineproject/gramine#1487] [General]
I have installed gramine. When I tried to run helloworld, I got the error below. I am not sure how to use gramine-sgx-get-token cmd.
Can anyone give me a hand? 

-------------------------
betrfs@bunsen20:~/authbekv/gramine/CI-Examples/helloworld$ gramine-sgx helloworld 

Gramine is starting. Parsing TOML manifest file, this may take some time...

error: Cannot open token helloworld.token. Use gramine-sgx-get-token on the runtime host to create the token file.

error: load_enclave() failed with error -22

betrfs@bunsen20:~/authbekv/gramine/CI-Examples/helloworld$

**Comment by kailun-qin** [comment]
Hi @jyizheng!

Just want to double check -- are you still using the SGX OOT driver? Whick kernel version do you have? Which Gramine version did you install? And how did you build/install it (perhaps by sharing us w/ the output of "cat `which gramine-sgx`")?

> I am not sure how to use gramine-sgx-get-token cmd. Can anyone give me a hand?

Pls kindly find the manpage of `gramine-sgx-get-token` here: https://gramine.readthedocs.io/en/latest/manpages/gramine-sgx-get-token.html.

But as indicated in the manpage: `Using this command is not necessary (it was previously), since the token is fetched automatically if needed during the first enclave start`, if you're using newer Gramine versions w/ the corresponding SGX driver correctly configured, this manual generation should not be needed (i.e., it's automatically done when you run `gramine-sgx`, pls see: https://github.com/gramineproject/gramine/blob/92636ae0bc8187d48dd40536d2a1901cb8eb6cd6/tools/gramine.in#L75-L86).

**Comment by jyizheng** [comment]
Thanks a lot of the reply. I am able to run helloworld now. Let me close this discussion.

**Comment by jyizheng** [comment]
I got an error below when running helloworld example in gramine. I have successfully installed gramine and SGX hardware is sgx1 with 128MB EPC.  Is there any idea what is going wrong according to the error message? 


(libos_vma.c:645:init_vma) debug: ASLR top address adjusted to 0x188c000
(libos_fs_encrypted.c:398:get_or_create_encrypted_files_key) debug: Successfully retrieved special key "_sgx_mrenclave"
(libos_fs_encrypted.c:398:get_or_create_encrypted_files_key) debug: Successfully retrieved special key "_sgx_mrsigner"
(attestation.c:365:init_sgx_attestation) debug: host is Linux-SGX and remote attestation type is 'none', skipping /dev/attestation/quote file
(libos_init.c:410:libos_init) debug: LibOS loaded at 0x3d57000, ready to initialize
(libos_init.c:561:create_pipe) debug: Creating pipe: pipe.srv:667266d4fe7c3d803b0613e04699c24e81a71be8acec4be744f9c1b901675e14
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "file:." (chroot) under /
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "proc" (pseudo) under /proc
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "dev" (pseudo) under /dev
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "dev:tty" (chroot) under /dev/tty
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "sys" (pseudo) under /sys
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "file:/usr/local/lib/x86_64-linux-gnu/gramine/runtime/glibc" (chroot) under /lib
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "file:helloworld" (chroot) under /helloworld
(libos_init.c:561:create_pipe) [P1:T1:helloworld] debug: Creating pipe: pipe.srv:c34f02c7e2f8fcae1e371b0a1ea64c85701410a67a40640eb593c69b3bea93b8
(libos_init.c:303:init_stack) [P1:T1:helloworld] debug: Allocated stack at 0x184c000 (size = 0x40000)
(libos_rtld.c:844:load_elf_object) [P1:T1:helloworld] debug: loading "file:helloworld"
(libos_rtld.c:898:find_interp) [P1:T1:helloworld] debug: searching for interpreter: /lib/ld-linux-x86-64.so.2
(libos_rtld.c:844:load_elf_object) [P1:T1:helloworld] debug: loading "file:/usr/local/lib/x86_64-linux-gnu/gramine/runtime/glibc/ld-linux-x86-64.so.2"
(libos_init.c:561:create_pipe) [P1:T1:helloworld] debug: Creating pipe: pipe.srv:9160656114594086635/1
(libos_init.c:561:create_pipe) [P1:T1:helloworld] debug: Creating pipe: pipe.srv:d9d1cc1ed7fa4ba755c3483e8edad157f48f544126261eb5e4d7b6385c0b562a
(libos_init.c:505:libos_init) [P1:T1:helloworld] debug: LibOS initialized
(libos_ipc_worker.c:372:ipc_worker_wrapper) [P1:libos] debug: IPC worker started
(libos_call.c:140:handle_libos_call) [P1:T1:helloworld] warning: handle_libos_call: invalid number: 25223248
(libos_signal.c:351:memfault_upcall) [P1:T1:helloworld] debug: memory fault at 0x00000000 (IP = 0x00000000)
(libos_signal.c:58:sighandler_kill) [P1:T1:helloworld] debug: killed by signal 11
(libos_fs_lock.c:785:file_lock_clear_pid) [P1:T1:helloworld] debug: clearing file (POSIX) locks for pid 1
(libos_sync_client.c:331:shutdown_sync_client) [P1:T1:helloworld] debug: sync client shutdown: closing handles
(libos_sync_client.c:346:shutdown_sync_client) [P1:T1:helloworld] debug: sync client shutdown: waiting for confirmation
(libos_sync_client.c:359:shutdown_sync_client) [P1:T1:helloworld] debug: sync client shutdown: finished
(libos_ipc_worker.c:285:ipc_worker_main) [P1:libos] debug: IPC worker: exiting worker thread
(libos_exit.c:58:libos_clean_and_exit) [P1:T1:helloworld] debug: process 1 exited with status 139
(pal_process.c:248:_PalProcessExit) debug: PalProcessExit: Returning exit code 139

**Reply by kailun-qin** [reply]
@jyizheng Can you pls set `loader.log_level = "trace"` in the manifest and show the log?

**Reply by jyizheng** [reply]
Sure. The log is below:

(libos_vma.c:645:init_vma) debug: ASLR top address adjusted to 0x46e6000
(libos_fs_encrypted.c:398:get_or_create_encrypted_files_key) debug: Successfully retrieved special key "_sgx_mrenclave"
(libos_fs_encrypted.c:398:get_or_create_encrypted_files_key) debug: Successfully retrieved special key "_sgx_mrsigner"
(attestation.c:365:init_sgx_attestation) debug: host is Linux-SGX and remote attestation type is 'none', skipping /dev/attestation/quote file
(libos_init.c:410:libos_init) debug: LibOS loaded at 0xfd57000, ready to initialize
(libos_init.c:561:create_pipe) debug: Creating pipe: pipe.srv:1f746e123c979f0cbb2a5191958c663caf3eef85aa5390b85b0e76b1d1612f21
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "file:." (chroot) under /
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "proc" (pseudo) under /proc
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "dev" (pseudo) under /dev
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "dev:tty" (chroot) under /dev/tty
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "sys" (pseudo) under /sys
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "file:/usr/local/lib/x86_64-linux-gnu/gramine/runtime/glibc" (chroot) under /lib
(libos_fs.c:525:mount_fs) [P1:T1:] debug: mounting "file:helloworld" (chroot) under /helloworld
(libos_init.c:561:create_pipe) [P1:T1:helloworld] debug: Creating pipe: pipe.srv:cb85dba360382e7ef6306f79b1307bc40dc7f22a01cee73f74fee1df3c4da3fc
(libos_init.c:303:init_stack) [P1:T1:helloworld] debug: Allocated stack at 0x46a6000 (size = 0x40000)
(libos_rtld.c:844:load_elf_object) [P1:T1:helloworld] debug: loading "file:helloworld"
(libos_rtld.c:898:find_interp) [P1:T1:helloworld] debug: searching for interpreter: /lib/ld-linux-x86-64.so.2
(libos_rtld.c:844:load_elf_object) [P1:T1:helloworld] debug: loading "file:/usr/local/lib/x86_64-linux-gnu/gramine/runtime/glibc/ld-linux-x86-64.so.2"
(libos_init.c:561:create_pipe) [P1:T1:helloworld] debug: Creating pipe: pipe.srv:2965200687339722861/1
(libos_init.c:561:create_pipe) [P1:T1:helloworld] debug: Creating pipe: pipe.srv:64cd263086f7496509752a609d1d8610ee317cb97da07828cd2644c492bc83a6
(libos_init.c:505:libos_init) [P1:T1:helloworld] debug: LibOS initialized
(libos_ipc_worker.c:372:ipc_worker_wrapper) [P1:libos] debug: IPC worker started
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- brk(0) = 0x4bdf000
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- mmap(0, 0x2000, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0x0) ...
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- return from mmap(...) = 0x4668000
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- access("/etc/ld.so.preload", F_OK|X_OK) = -2
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- openat(AT_FDCWD, "/lib/glibc-hwcaps/x86-64-v3/libc.so.6", O_RDONLY|0x80000, 0000) = -2
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- newfstatat(AT_FDCWD, "/lib/glibc-hwcaps/x86-64-v3", 0x46e4f70, 0) = -2
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- openat(AT_FDCWD, "/lib/glibc-hwcaps/x86-64-v2/libc.so.6", O_RDONLY|0x80000, 0000) = -2
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- newfstatat(AT_FDCWD, "/lib/glibc-hwcaps/x86-64-v2", 0x46e4f70, 0) = -2
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- openat(AT_FDCWD, "/lib/libc.so.6", O_RDONLY|0x80000, 0000) = 0x3
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- read(3, 0x46e50e8, 0x340) ...
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- return from read(...) = 0x340
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- newfstatat(3, "", 0x46e4f70, 4096) = 0x0
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- mmap(0, 0x1cfd30, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0x0) ...
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- return from mmap(...) = 0x4498000
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- mmap(0x465a000, 0x6000, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1c1000) ...
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- return from mmap(...) = 0x465a000
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- mmap(0x4660000, 0x7d30, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED, -1, 0x0) ...
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- return from mmap(...) = 0x4660000
(libos_parser.c:1628:buf_write_all) [P1:T1:helloworld] trace: ---- close(3) = 0x0
(libos_call.c:140:handle_libos_call) [P1:T1:helloworld] warning: handle_libos_call: invalid number: 73826384
(libos_signal.c:351:memfault_upcall) [P1:T1:helloworld] debug: memory fault at 0x00000000 (IP = 0x00000000)
(libos_signal.c:58:sighandler_kill) [P1:T1:helloworld] debug: killed by signal 11
(libos_fs_lock.c:785:file_lock_clear_pid) [P1:T1:helloworld] debug: clearing file (POSIX) locks for pid 1
(libos_sync_client.c:331:shutdown_sync_client) [P1:T1:helloworld] debug: sync client shutdown: closing handles
(libos_sync_client.c:346:shutdown_sync_client) [P1:T1:helloworld] debug: sync client shutdown: waiting for confirmation
(libos_sync_client.c:359:shutdown_sync_client) [P1:T1:helloworld] debug: sync client shutdown: finished
(libos_ipc_worker.c:285:ipc_worker_main) [P1:libos] debug: IPC worker: exiting worker thread
(libos_exit.c:58:libos_clean_and_exit) [P1:T1:helloworld] debug: process 1 exited with status 139
(pal_process.c:248:_PalProcessExit) debug: PalProcessExit: Returning exit code 139

**Reply by kailun-qin** [reply]
It looks like it failed at `gramine_register_library()` when dynamic loading `libc.so`: https://github.com/gramineproject/gramine/blob/92636ae0bc8187d48dd40536d2a1901cb8eb6cd6/libos/include/arch/x86_64/gramine_entry_api.h#L78, which was invoked by our patched glibc (pls see: https://github.com/gramineproject/gramine/blob/92636ae0bc8187d48dd40536d2a1901cb8eb6cd6/subprojects/packagefiles/glibc-2.37/gramine-syscall.patch#L53).

However w/o a reproduction env, sorry IDK how this happened. I'd suggest to:
1. Redo a clean build of Gramine (e.g., by removing first `/usr/local/lib/x86_64-linux-gnu/gramine`).
2. If the issue persists, pls share your kernel, distro and compiler versions w/ us.
3. If the issue persists, consider to confirm it by GDB into `gramine-sgx helloworld`, break pointing at `gramine_call()`, and checking the related info of registers.
4. If the issue persists, check the related inline functions produced in the binary of Gramine-generated `ld.so` and see if anything suspicious.

**Comment by jyizheng** [comment]
I can make helloworld work after switching to  gcc-9. 

I also find a bug in the build script.

https://github.com/gramineproject/gramine/blob/92636ae0bc8187d48dd40536d2a1901cb8eb6cd6/subprojects/packagefiles/glibc-2.37/compile.sh#L22C1-L23C1

 -Wno-unused-values  -->  -Wno-unused-value

**Reply by kailun-qin** [reply]
> I can make helloworld work after switching to gcc-9.

May I know the compiler version where you met the original issue?

> I also find a bug in the build script.

Thanks a lot for the catch! I submmited https://github.com/gramineproject/gramine/pull/1490 w/ a "Reported-by" tag.

**Reply by jyizheng** [reply]
It is gcc 7.5

**Comment by jyizheng** [comment]
I rebuilt gramine from the latest master branch. Then I tried to run helloworld. I got an error below.

error: Unrecognized dynamic entry (DT_*) 24
error: Relocation of the PAL binary failed: -6
 
Is there any idea what is going wrong?

**Reply by kailun-qin** [reply]
This should be caused by the recently introduced PR: https://github.com/gramineproject/gramine/pull/1473 where we fail if Gramine binaries have unrecognized or unexpected/malformed dyn entries. Looks like the `DT_BIND_NOW` was found in your built Gramine binaries.

May I know your current compiler version? Thanks!

cc @dimakuv

**Reply by dimakuv** [reply]
@jyizheng Interesting problem with DT 24, which is [`DT_BIND_NOW`](https://github.com/gramineproject/gramine/blob/a8edb2e17d7b9e95ff46e8ee4dcee47447808f33/pal/include/elf/elf.h#L652).

Please state your compiler version, your OS distro, and the way you built Gramine.

It looks like some compilers use `DT_BIND_NOW`, though this particular dynentry was superseded by a flag `DF_BIND_NOW`. At the same time, this dynentry simply means "use immediate binding instead of lazy binding" (lazy binding is the default for linkers). And our Gramine-internal linker always uses immediate binding. So we can simply ignore `DT_BIND_NOW` dynentry, as it just hints at what Gramine is doing anyway.

References:
- http://www.qnx.com/developers/docs/qnxcar2/index.jsp?topic=%2Fcom.qnx.doc.neutrino.prog%2Ftopic%2Fdevel_Lazy_binding.html
- https://github.com/gramineproject/gramine/blob/a8edb2e17d7b9e95ff46e8ee4dcee47447808f33/libos/src/meson.build#L159?plain=1
- https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-42444.html
- https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter3-29.html#chapter3-4

Let me cook a quick PR that whitelists `DT_BIND_NOW`.

**Reply by dimakuv** [reply]
@jyizheng @kailun-qin I created a PR: https://github.com/gramineproject/gramine/pull/1508.

I'll be glad if @jyizheng could check whether this PR solves his problem of `error: Unrecognized dynamic entry (DT_*) 24`.
